man/boobtracker.1
.\" -*- coding: utf-8 -*-
.\" This file was generated automatically by tools/make_man.sh.
.TH BOOBTRACKER 1 "13 February 2020" "boobtracker 2\&.0"
.SH NAME
boobtracker \- manage bug tracking issues
.SH SYNOPSIS
.B boobtracker
[\-dqv] [\-b \fIbackends\fR] [\-cnfs] [\fIcommand\fR [\fIarguments\fR..]]
.br
.B boobtracker
[\-\-help] [\-\-version]

.SH DESCRIPTION
.LP

Console application allowing to create, edit, view bug tracking issues.

.SS Supported websites:
* asana (Asana)
.br
* github (GitHub issues tracking)
.br
* redmine (The Redmine project management web application)
.SH BOOBTRACKER COMMANDS
.TP
\fBattach\fR \fIISSUE\fR \fIFILENAME\fR
.br
Attach a file to an issue (Not implemented yet).
.TP
\fBcomment\fR \fIISSUE\fR [\fITEXT\fR]
.br
Comment an issue. If no text is given, enter it in standard input.
.TP
\fBdebug\fR
.br
Launch a debug Python shell
.TP
\fBedit\fR \fIISSUE\fR [\fIKEY\fR [\fIVALUE\fR]]
.br
Edit an issue.
.br
If you are not in interactive mode, you can use these parameters:
.br
\-\-title TITLE
.br
\-\-assignee ASSIGNEE
.br
\-\-target\-version VERSION
.br
\-\-category CATEGORY
.br
\-\-status STATUS
.TP
\fBget\fR \fIISSUE\fR
.br
Get an issue and display it.
.TP
\fBlogtime\fR \fIISSUE\fR \fIHOURS\fR [\fITEXT\fR]
.br
Log spent time on an issue.
.TP
\fBpost\fR \fIPROJECT\fR
.br
Post a new issue.
.br

.br
If you are not in interactive mode, you can use these parameters:
.br
\-\-title TITLE
.br
\-\-assignee ASSIGNEE
.br
\-\-target\-version VERSION
.br
\-\-category CATEGORY
.br
\-\-status STATUS
.TP
\fBremove\fR \fIISSUE\fR
.br
Remove an issue.
.TP
\fBsearch\fR \fIPROJECT\fR
.br
List issues for a project.
.br

.br
You can use these filters from command line:
.br
\-\-author AUTHOR
.br
\-\-title TITLE_PATTERN
.br
\-\-assignee ASSIGNEE
.br
\-\-target\-version VERSION
.br
\-\-category CATEGORY
.br
\-\-status STATUS
.br

.br
Default is limited to 10 results.
.SH WEBOOB COMMANDS
.TP
\fBbackends\fR [\fIACTION\fR] [\fIBACKEND_NAME\fR]...
.br
Select used backends.
.br

.br
ACTION is one of the following (default: list):
.br
* enable         enable given backends
.br
* disable        disable given backends
.br
* only           enable given backends and disable the others
.br
* list           list backends
.br
* add            add a backend
.br
* register       register a new account on a website
.br
* edit           edit a backend
.br
* remove         remove a backend
.br
* list\-modules   list modules
.TP
\fBcd\fR [\fIPATH\fR]
.br
Follow a path.
.br
".." is a special case and goes up one directory.
.br
"" is a special case and goes home.
.TP
\fBcondition\fR [\fIEXPRESSION\fR | off]
.br
If an argument is given, set the condition expression used to filter the results. See CONDITION section for more details and the expression.
.br
If the "off" value is given, conditional filtering is disabled.
.br

.br
If no argument is given, print the current condition expression.
.TP
\fBcount\fR [\fINUMBER\fR | off]
.br
If an argument is given, set the maximum number of results fetched.
.br
NUMBER must be at least 1.
.br
"off" value disables counting, and allows infinite searches.
.br

.br
If no argument is given, print the current count value.
.TP
\fBformatter\fR [list | \fIFORMATTER\fR [\fICOMMAND\fR] | option \fIOPTION_NAME\fR [on | off]]
.br
If a FORMATTER is given, set the formatter to use.
.br
You can add a COMMAND to apply the formatter change only to
.br
a given command.
.br

.br
If the argument is "list", print the available formatters.
.br

.br
If the argument is "option", set the formatter options.
.br
Valid options are: header, keys.
.br
If on/off value is given, set the value of the option.
.br
If not, print the current value for the option.
.br

.br
If no argument is given, print the current formatter.
.TP
\fBlogging\fR [\fILEVEL\fR]
.br
Set logging level.
.br

.br
Availables: debug, info, warning, error.
.br
* quiet is an alias for error
.br
* default is an alias for warning
.TP
\fBls\fR [\-d] [\-\fIU\fR] [\fIPATH\fR]
.br
List objects in current path.
.br
If an argument is given, list the specified path.
.br
Use \-U option to not sort results. It allows you to use a "fast path" to
.br
return results as soon as possible.
.br
Use \-d option to display information about a collection (and to not
.br
display the content of it). It has the same behavior than the well
.br
known UNIX "ls" command.
.br

.br
Default is limited to 40 results.
.TP
\fBquit\fR
.br
Quit the application.
.TP
\fBselect\fR [\fIFIELD_NAME\fR]... | "$direct" | "$full"
.br
If an argument is given, set the selected fields.
.br
$direct selects all fields loaded in one http request.
.br
$full selects all fields using as much http requests as necessary.
.br

.br
If no argument is given, print the currently selected fields.

.SH OPTIONS
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-b BACKENDS\fR, \fB\-\-backends=BACKENDS\fR
what backend(s) to enable (comma separated)
.TP
\fB\-e EXCLUDE_BACKENDS\fR, \fB\-\-exclude\-backends=EXCLUDE_BACKENDS\fR
what backend(s) to exclude (comma separated)
.TP
\fB\-I\fR, \fB\-\-insecure\fR
do not validate SSL
.TP
\fB\-\-nss\fR
Use NSS instead of OpenSSL
.TP
\fB\-\-auto\-update\fR
Automatically check for updates when a bug in a module is encountered

.SH BOOBTRACKER OPTIONS
.TP
\fB\-\-author=AUTHOR\fR
.TP
\fB\-\-title=TITLE\fR
.TP
\fB\-\-assignee=ASSIGNEE\fR
.TP
\fB\-\-target\-version=VERSION\fR
.TP
\fB\-\-tracker=TRACKER\fR
.TP
\fB\-\-category=CATEGORY\fR
.TP
\fB\-\-status=STATUS\fR
.TP
\fB\-\-priority=PRIORITY\fR
.TP
\fB\-\-start=START\fR
.TP
\fB\-\-due=DUE\fR

.SH LOGGING OPTIONS
.TP
\fB\-d\fR, \fB\-\-debug\fR
display debug messages. Set up it twice to more verbosity
.TP
\fB\-q\fR, \fB\-\-quiet\fR
display only error messages
.TP
\fB\-v\fR, \fB\-\-verbose\fR
display info messages
.TP
\fB\-\-logging\-file=LOGGING_FILE\fR
file to save logs
.TP
\fB\-a\fR, \fB\-\-save\-responses\fR
save every response
.TP
\fB\-\-export\-session\fR
log browser session cookies after login

.SH RESULTS OPTIONS
.TP
\fB\-c CONDITION\fR, \fB\-\-condition=CONDITION\fR
filter result items to display given a boolean expression. See CONDITION section
for the syntax
.TP
\fB\-n COUNT\fR, \fB\-\-count=COUNT\fR
limit number of results (from each backends)
.TP
\fB\-s SELECT\fR, \fB\-\-select=SELECT\fR
select result item keys to display (comma separated)

.SH FORMATTING OPTIONS
.TP
\fB\-f FORMATTER\fR, \fB\-\-formatter=FORMATTER\fR
select output formatter (csv, htmltable, issue_info, issues_list, json,
json_line, multiline, simple, table, webkit)
.TP
\fB\-\-no\-header\fR
do not display header
.TP
\fB\-\-no\-keys\fR
do not display item keys
.TP
\fB\-O OUTFILE\fR, \fB\-\-outfile=OUTFILE\fR
file to export result

.SH CONDITION
The \-c and \-\-condition is a flexible way to filter and get only interesting results. It supports conditions on numerical values, dates, durations and strings. Dates are given in YYYY\-MM\-DD or YYYY\-MM\-DD HH:MM format. Durations look like XhYmZs where X, Y and Z are integers. Any of them may be omitted. For instance, YmZs, XhZs or Ym are accepted.
The syntax of one expression is "\fBfield operator value\fR". The field to test is always the left member of the expression.
.LP
The field is a member of the objects returned by the command. For example, a bank account has "balance", "coming" or "label" fields.
.SS The following operators are supported:
.TP
=
Test if object.field is equal to the value.
.TP
!=
Test if object.field is not equal to the value.
.TP
>
Test if object.field is greater than the value. If object.field is date, return true if value is before that object.field.
.TP
<
Test if object.field is less than the value. If object.field is date, return true if value is after that object.field.
.TP
|
This operator is available only for string fields. It works like the Unix standard \fBgrep\fR command, and returns True if the pattern specified in the value is in object.field.
.SS Expression combination
.LP
You can make a expression combinations with the keywords \fB" AND "\fR, \fB" OR "\fR an \fB" LIMIT "\fR.
.LP
The \fBLIMIT\fR keyword can be used to limit the number of items upon which running the expression. \fBLIMIT\fR can only be placed at the end of the expression followed by the number of elements you want.
.SS Examples:
.nf
.B boobank ls \-\-condition 'label=Livret A'
.fi
Display only the "Livret A" account.
.PP
.nf
.B boobank ls \-\-condition 'balance>10000'
.fi
Display accounts with a lot of money.
.PP
.nf
.B boobank history account@backend \-\-condition 'label|rewe'
.fi
Get transactions containing "rewe".
.PP
.nf
.B boobank history account@backend \-\-condition 'date>2013\-12\-01 AND date<2013\-12\-09'
.fi
Get transactions betweens the 2th December and 8th December 2013.
.PP
.nf
.B boobank history account@backend \-\-condition 'date>2013\-12\-01  LIMIT 10'
.fi
Get transactions after the 2th December in the last 10 transactions

.SH COPYRIGHT
Copyright(C) 2011-2020 Romain Bignon
.LP
For full copyright information see the COPYING file in the weboob package.
.LP
.RE
.SH FILES
"~/.config/weboob/backends" 

.SH SEE ALSO
Home page: http://weboob.org/applications/boobtracker


ChangeLog
Weboob 2.0 (2020-02-13)

	Applications: boobank
	* keep Value for TransferStep and AddRecipientStep
	* declare utf-8 in OFX formatter
	* give account object to iter_transfer_recipient

	Applications: boobill
	* using a persistent storage for browser state

	Applications: boobsize
	* fix name

	Applications: translaboob
	* use babel to detect language names and pass ISO-639 codes

	Applications: weboobdebug
	* weboob-debug: fix name

	Applications: weboobmain
	* move scripts/weboob to a dedicated dir in weboob/applications

	Applications: weboobrepos
	* weboob-repos: port to python3

	Browser
	* set a parent logger to AbstractElement
	* Allows SeleniumBrowser to use MANUAL proxies
	* AUTHENTICATION_METHODS set in init
	* delete prefix 'handle_' on authentication methods
	* clean authentication config keys after login
	* make variable name more understandable
	* make get_expire stronger
	* TWOFA_DURATION for the state expire date
	* better names
	* set expire from logged_date
	* make authentication methods more flexible
	* handle both 2FA and regular login methods
	* add possibility to clear specific cookies
	* handle NeedInteractiveFor2FA
	* add documentation about otp, sms and resume
	* add TwoFactorBrowser to handle multiple 2FA
	* Set default logging level to ERROR for selenium
	* Added the ability to overrride *Elements' env copy
	* Allow passing custom headers in Form.submit()
	* set attribute 'weboob' on browser
	* Normalize encodings
	* BrowserTooManyRequests: Remove class attributes for date helpers
	* BrowserTooManyRequests: Raise a TypeError if the next_try is not a datetime
	* BrowserTooManyRequests: Check whether the next_try is None after the date check
	* BrowserTooManyRequests: Add message parameter
	* [weboob/browser/browser+pages] Fixed doctest BASEURL
	* add some helpers to TooManyRequests
	* add rate limiting exception
	* add build_refresh_token_parameters method
	* python3 compatibility
	* send redirect_uri when using the refresh_token
	* allow multiple inheritance with AbstractBrowser
	* Implemented wait_xpath_invisible
	* Avoid crash when no BASEURL is set
	* add klass attribute to decorated @method
	* give state to the module at initialization
	* introduce BrowserParamURL for passing params implictly
	* Implement the empty_xpath attribute
	* accept new parameter proxy_headers (but don't use it)
	* Browser takes a new parameter 'proxy_headers'
	* if use of refresh_token fails, forget it
	* support when token_type key is missing
	* if access_token_expire is None, the access_token has no expiration
	* in case we got a 401 response, forget the access_token
	* fix compatibility with python3
	* Handle generic error for bad url callback
	* Repair refresh token
	* StatesMixin cookies should be str
	* OAuth2PKCEMixin codes shouldn't be bytes
	* Rewrite/improve PagesBrowser example

	Browser: Filters
	* Add the format .Italian() to CleanDecimal
	* New keyword argument to remove accented characters on CleanText filter.
	* Changed Capitalize filter to Title
	* Updated & Reordered __all__
	* Add dates to FormValue
	* Add color to FormValue
	* Add number/range type to FormValue, and tests
	* MultiJoin accepts types other than str
	* Regexp can use regex library
	* Allow all dateutil.parser arguments in DateTime
	* Strict dates by default
	* Catch incomplete dates
	* Explicit error message
	* Add MultiJoin
	* Join can use default argument
	* Add doctests to Join
	* TableCell(colspan=True) is now the only behavior desired
	* add FromTimestamp to convert to datetime
	* Fix error message for FormValue
	* Add new types for FormValue
	* Add MapIn filter, that allow the pattern of a selected value to another value using a dict.
	* Reorganize imports / add missing to __all__
	* Add Coalesce

	Capabilities
	* rename accepted_doc_types attribute of CapDocument
	* Add KIID to DocumentTypes
	* Added new capability CapBands
	* Keep only PER
	* add PER account type
	* HousingPhoto inherits BaseImage
	* use PostalAddress and GeoCoords types where possible
	* introduce weboob.capabilities.address for a few object types
	* Add firstname and lastname field
	* add capability_to_string function to get capability name
	* Add new SAR currency
	* fix missing TransferInvalidOTP export
	* add TransferInvalidOTP exception
	* use StringField standard default value for transfer beneficiary label and number
	* Indentation cosmetic changes
	* Add new Investment attributes
	* Add new document type 'income_tax'
	* Add field 'original_gross_amount' to Transaction
	* Add field 'gross_amount' to Transaction
	* Remove performance_history default value
	* Added new attributes for employee savings accounts
	* Add RecaptchaV3Job to captcha capability
	* add support for commission in currency in bank/Transaction
	* add new add recipient exception for invalid otp
	* add has_file field to Document class
	* Add the relationship between the credentials owner (PSU) and the account
	* Add bdate. Represent the date that the transaction appear in the website.
	* prevent casting id from None to 'None' in constructor

	Documentation
	* add pip install
	* fix compatibility with sphinx 2.0
	* do not use deprecated defindex.html anymore
	* Use correct word

	General
	* New airparif module (CapGauge)
	* New aviva module (CapBank, CapBankWealth)
	* New cesu module (CapDocument)
	* New degiro module (CapBank, CapBankWealth)
	* New hsbchk module (CapBank)
	* New lendosphere module (CapBank, CapBankWealth)
	* New lucca module (CapCalendarEvent, CapDocument)
	* New lunchr module (CapBank)
	* New metalarchives module (CapBands)
	* New pajemploi module (CapDocument)
	* New ticketscesu module (CapBank)
	* New wiseed module (CapBank, CapBankWealth)
	* Deleted biplan module
	* Deleted figgo module
	* Deleted lolix module
	* Deleted wellsfargo module

	General: Core
	* send request_information to modules in all case, except if stdout is not a tty
	* handle error NeedInteractiveFor2FA

weboob/applications/boobtracker/boobtracker.py
# -*- coding: utf-8 -*-

# Copyright(C) 2011  Romain Bignon
#
# This file is part of weboob.
#
# weboob is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# weboob is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with weboob. If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

from datetime import timedelta
from email import message_from_string, message_from_file
from email.header import decode_header
from email.mime.text import MIMEText
from smtplib import SMTP
import os
import re
import unicodedata

from weboob.capabilities.base import empty, BaseObject
from weboob.capabilities.bugtracker import CapBugTracker, Query, Update, Project, Issue, IssueError
from weboob.tools.application.repl import ReplApplication, defaultcount
from weboob.tools.application.formatters.iformatter import IFormatter, PrettyFormatter
from weboob.tools.compat import basestring, unicode
from weboob.tools.html import html2text
from weboob.tools.date import parse_french_date


__all__ = ['BoobTracker']


try:
    input = raw_input
except NameError:
    pass


class IssueFormatter(IFormatter):
    MANDATORY_FIELDS = ('id', 'project', 'title', 'body', 'author')

    def format_attr(self, obj, attr):
        if not hasattr(obj, attr) or empty(getattr(obj, attr)):
            return u''

        value = getattr(obj, attr)
        if isinstance(value, BaseObject):
            value = value.name

        return self.format_key(attr.capitalize(), value)

    def format_key(self, key, value):
        return '%s %s\n' % (self.colored('%s:' % key, 'green'),
                            value)

    def format_obj(self, obj, alias):
        result = u'%s %s %s %s %s\n' % (self.colored(obj.project.name, 'blue', 'bold'),
                                        self.colored(u'—', 'cyan', 'bold'),
                                        self.colored(obj.fullid, 'red', 'bold'),
                                        self.colored(u'—', 'cyan', 'bold'),
                                        self.colored(obj.title, 'yellow', 'bold'))
        result += '\n%s\n\n' % obj.body
        result += self.format_key('Author', '%s (%s)' % (obj.author.name, obj.creation))
        result += self.format_attr(obj, 'status')
        result += self.format_attr(obj, 'priority')
        result += self.format_attr(obj, 'version')
        result += self.format_attr(obj, 'tracker')
        result += self.format_attr(obj, 'category')
        result += self.format_attr(obj, 'assignee')
        if hasattr(obj, 'fields') and not empty(obj.fields):
            for key, value in obj.fields.items():
                result += self.format_key(key.capitalize(), value)
        if hasattr(obj, 'attachments') and obj.attachments:
            result += '\n%s\n' % self.colored('Attachments:', 'green')
            for a in obj.attachments:
                result += '* %s%s%s <%s>\n' % (self.BOLD, a.filename, self.NC, a.url)
        if hasattr(obj, 'history') and obj.history:
            result += '\n%s\n' % self.colored('History:', 'green')
            for u in obj.history:
                result += '%s %s %s %s\n' % (self.colored('*', 'red', 'bold'),
                                             self.colored(u.date, 'yellow', 'bold'),
                                             self.colored(u'—', 'cyan', 'bold'),
                                             self.colored(u.author.name, 'blue', 'bold'))
                for change in u.changes:
                    result += '  - %s %s %s %s\n' % (self.colored(change.field, 'green'),
                                                     change.last,
                                                     self.colored('->', 'magenta'), change.new)
                if u.message:
                    result += '    %s\n' % html2text(u.message).strip().replace('\n', '\n    ')
        return result


class IssuesListFormatter(PrettyFormatter):
    MANDATORY_FIELDS = ('id', 'project', 'status', 'title', 'category')

    def get_title(self, obj):
        return '%s - [%s] %s' % (obj.project.name, obj.status.name, obj.title)

    def get_description(self, obj):
        return obj.category


class BoobTracker(ReplApplication):
    APPNAME = 'boobtracker'
    VERSION = '2.1'
    COPYRIGHT = 'Copyright(C) 2011-YEAR Romain Bignon'
    DESCRIPTION = "Console application allowing to create, edit, view bug tracking issues."
    SHORT_DESCRIPTION = "manage bug tracking issues"
    CAPS = CapBugTracker
    EXTRA_FORMATTERS = {'issue_info': IssueFormatter,
                        'issues_list': IssuesListFormatter,
                       }
    COMMANDS_FORMATTERS = {'get':     'issue_info',
                           'post':    'issue_info',
                           'edit':    'issue_info',
                           'search':  'issues_list',
                           'ls':      'issues_list',
                          }
    COLLECTION_OBJECTS = (Project, Issue, )

    def add_application_options(self, group):
        group.add_option('--author')
        group.add_option('--title')
        group.add_option('--assignee')
        group.add_option('--target-version', dest='version')
        group.add_option('--tracker')
        group.add_option('--category')
        group.add_option('--status')
        group.add_option('--priority')
        group.add_option('--start')
        group.add_option('--due')

    @defaultcount(10)
    def do_search(self, line):
        """
        search PROJECT

        List issues for a project.

        You can use these filters from command line:
           --author AUTHOR
           --title TITLE_PATTERN
           --assignee ASSIGNEE
           --target-version VERSION
           --category CATEGORY
           --status STATUS
        """
        query = Query()

        path = self.working_path.get()
        backends = []
        if line.strip():
            query.project, backends = self.parse_id(line, unique_backend=True)
        elif len(path) > 0:
            query.project = path[0]
        else:
            print('Please enter a project name', file=self.stderr)
            return 1

        query.author = self.options.author
        query.title = self.options.title
        query.assignee = self.options.assignee
        query.version = self.options.version
        query.category = self.options.category
        query.status = self.options.status

        self.change_path([query.project, u'search'])
        for issue in self.do('iter_issues', query, backends=backends):
            self.add_object(issue)
            self.format(issue)

    def complete_get(self, text, line, *ignored):
        args = line.split(' ')
        if len(args) == 2:
            return self._complete_object()

    def do_get(self, line):
        """
        get ISSUE

        Get an issue and display it.
        """
        if not line:
            print('This command takes an argument: %s' % self.get_command_help('get', short=True), file=self.stderr)
            return 2

        issue = self.get_object(line, 'get_issue')
        if not issue:
            print('Issue not found: %s' % line, file=self.stderr)
            return 3
        self.format(issue)

    def complete_comment(self, text, line, *ignored):
        args = line.split(' ')
        if len(args) == 2:
            return self._complete_object()

    def do_comment(self, line):
        """
        comment ISSUE [TEXT]

        Comment an issue. If no text is given, enter it in standard input.
        """
        id, text = self.parse_command_args(line, 2, 1)
        if text is None:
            text = self.acquire_input()

        id, backend_name = self.parse_id(id, unique_backend=True)
        update = Update(0)
        update.message = text

        self.do('update_issue', id, update, backends=backend_name).wait()

    def do_logtime(self, line):
        """
        logtime ISSUE HOURS [TEXT]

        Log spent time on an issue.
        """
        id, hours, text = self.parse_command_args(line, 3, 2)
        if text is None:
            text = self.acquire_input()

        try:
            hours = float(hours)
        except ValueError:
            print('Error: HOURS parameter may be a float', file=self.stderr)
            return 1

        id, backend_name = self.parse_id(id, unique_backend=True)
        update = Update(0)
        update.message = text
        update.hours = timedelta(hours=hours)

        self.do('update_issue', id, update, backends=backend_name).wait()

    def complete_remove(self, text, line, *ignored):
        args = line.split(' ')
        if len(args) == 2:
            return self._complete_object()

    def do_remove(self, line):
        """
        remove ISSUE

        Remove an issue.
        """
        id, backend_name = self.parse_id(line, unique_backend=True)
        self.do('remove_issue', id, backends=backend_name).wait()

    ISSUE_FIELDS = (('title',    (None,       False)),
                    ('assignee', ('members',  True)),
                    ('version',  ('versions', True)),
                    ('tracker',  (None,       False)),#XXX
                    ('category', ('categories', False)),
                    ('status',   ('statuses', True)),
                    ('priority', (None,       False)),#XXX
                    ('start',    (None,       False)),
                    ('due',      (None,       False)),
                   )

    def get_list_item(self, objects_list, name):
        if name is None:
            return None

        for obj in objects_list:
            if obj.name.lower() == name.lower():
                return obj

        if not name:
            return None

        raise ValueError('"%s" is not found' % name)

    def sanitize_key(self, key):
        if isinstance(key, str):
            key = unicode(key, "utf8")
        key = unicodedata.normalize('NFKD', key).encode("ascii", "ignore")
        return key.replace(' ', '-').capitalize()

    def issue2text(self, issue, backend=None):
        if backend is not None and 'username' in backend.config:
            sender = backend.config['username'].get()
        else:
            sender = os.environ.get('USERNAME', 'boobtracker')
        output = u'From: %s\n' % sender
        for key, (list_name, is_list_object) in self.ISSUE_FIELDS:
            value = None
            if not self.interactive:
                value = getattr(self.options, key)
            if not value:
                value = getattr(issue, key)
            if not value:
                value = ''
            elif hasattr(value, 'name'):
                value = value.name

            if list_name is not None:
                objects_list = getattr(issue.project, list_name)
                if len(objects_list) == 0:
                    continue

            output += '%s: %s\n' % (self.sanitize_key(key), value)
            if list_name is not None:
                availables = ', '.join(['<%s>' % (o if isinstance(o, basestring) else o.name)
                                        for o in objects_list])
                output += 'X-Available-%s: %s\n' % (self.sanitize_key(key), availables)

        for key, value in issue.fields.items():
            output += '%s: %s\n' % (self.sanitize_key(key), value or '')
            # TODO: Add X-Available-* for lists

        output += '\n%s' % (issue.body or 'Please write your bug report here.')
        return output

    def text2issue(self, issue, m):
        # XXX HACK to support real incoming emails
        if 'Subject' in m:
            m['Title'] = m['Subject']

        for key, (list_name, is_list_object) in self.ISSUE_FIELDS:
            value = m.get(key)
            if value is None:
                continue

            new_value = u''
            for part in decode_header(value):
                if part[1]:
                    new_value += unicode(part[0], part[1])
                else:
                    new_value += part[0].decode('utf-8')
            value = new_value

            if is_list_object:
                objects_list = getattr(issue.project, list_name)
                value = self.get_list_item(objects_list, value)

            # FIXME: autodetect
            if key in ['start', 'due']:
                if len(value) > 0:
                    #value = datetime.strptime(value, "%Y-%m-%d %H:%M:%S")
                    value = parse_french_date(value)
                else:
                    value = None

            setattr(issue, key, value)

        for key in issue.fields.keys():
            value = m.get(self.sanitize_key(key))
            if value is not None:
                issue.fields[key] = value.decode('utf-8')

        content = u''
        for part in m.walk():
            if part.get_content_type() == 'text/plain':
                s = part.get_payload(decode=True)
                charsets = part.get_charsets() + m.get_charsets()
                for charset in charsets:
                    try:
                        if charset is not None:
                            content += unicode(s, charset)
                        else:
                            content += unicode(s, encoding='utf-8')
                    except UnicodeError as e:
                        self.logger.warning('Unicode error: %s' % e)
                        continue
                    except Exception as e:
                        self.logger.exception(e)
                        continue
                    else:
                        break

        issue.body = content

        m = re.search('([^< ]+@[^ >]+)', m['From'] or '')
        if m:
            return m.group(1)

    def edit_issue(self, issue, edit=True):
        backend = self.weboob.get_backend(issue.backend)
        content = self.issue2text(issue, backend)
        while True:
            if self.stdin.isatty():
                content = self.acquire_input(content, {'vim': "-c 'set ft=mail'"})
                m = message_from_string(content.encode('utf-8'))
            else:
                m = message_from_file(self.stdin)

            try:
                email_to = self.text2issue(issue, m)
            except ValueError as e:
                if not self.stdin.isatty():
                    raise
                input("%s -- Press Enter to continue..." % unicode(e).encode("utf-8"))
                continue

            try:
                issue = backend.post_issue(issue)
                print('Issue %s %s' % (self.formatter.colored(issue.fullid, 'red', 'bold'),
                                       'updated' if edit else 'created'))
                if edit:
                    self.format(issue)
                elif email_to:
                    self.send_notification(email_to, issue)
                return 0
            except IssueError as e:
                if not self.stdin.isatty():
                    raise
                input("%s -- Press Enter to continue..." % unicode(e).encode("utf-8"))

    def send_notification(self, email_to, issue):
        text = """Hi,

You have successfuly created this ticket on the Weboob tracker:

%s

You can follow your bug report on this page:

https://symlink.me/issues/%s

Regards,

Weboob Team
""" % (issue.title, issue.id)
        msg = MIMEText(text, 'plain', 'utf-8')
        msg['Subject'] = 'Issue #%s reported' % issue.id
        msg['From'] = 'Weboob <weboob@weboob.org>'
        msg['To'] = email_to
        s = SMTP('localhost')
        s.sendmail('weboob@weboob.org', [email_to], msg.as_string())
        s.quit()

    def do_post(self, line):
        """
        post PROJECT

        Post a new issue.

        If you are not in interactive mode, you can use these parameters:
           --title TITLE
           --assignee ASSIGNEE
           --target-version VERSION
           --category CATEGORY
           --status STATUS
        """
        if not line.strip():
            print('Please give the project name')
            return 1

        project, backend_name = self.parse_id(line, unique_backend=True)

        backend = self.weboob.get_backend(backend_name)

        issue = backend.create_issue(project)
        issue.backend = backend.name

        return self.edit_issue(issue, edit=False)

    def complete_edit(self, text, line, *ignored):
        args = line.split(' ')
        if len(args) == 2:
            return self._complete_object()
        if len(args) == 3:
            return list(dict(self.ISSUE_FIELDS).keys())

    def do_edit(self, line):
        """
        edit ISSUE [KEY [VALUE]]

        Edit an issue.
        If you are not in interactive mode, you can use these parameters:
           --title TITLE
           --assignee ASSIGNEE
           --target-version VERSION
           --category CATEGORY
           --status STATUS
        """
        _id, key, value = self.parse_command_args(line, 3, 1)
        issue = self.get_object(_id, 'get_issue')
        if not issue:
            print('Issue not found: %s' % _id, file=self.stderr)
            return 3

        return self.edit_issue(issue, edit=True)

    def complete_attach(self, text, line, *ignored):
        args = line.split(' ')
        if len(args) == 2:
            return self._complete_object()
        elif len(args) >= 3:
            return self.path_completer(args[2])

    def do_attach(self, line):
        """
        attach ISSUE FILENAME

        Attach a file to an issue (Not implemented yet).
        """
        print('Not implemented yet.', file=self.stderr)
